### PR TITLE
Vscode devcontainer

### PR DESCRIPTION
**Pull request by bhack**
I know we closed https://github.com/tensorflow/tensorflow/pull/48679 but as Codespaces is GA now please keep this open so that in the meantime we have a PR where the user could test and bootstrap Github Codespaces and we could collect some feedback.

### CODE REVIEW

**Comment by bhack** [comment]
Keras has already merged its own https://github.com/keras-team/keras/commit/4c9cf436ca9b1b3a1e8bfd0c89dc96ec855f2233

**Comment by bhack** [comment]
I don't know if Tensorflow team members have this setting enabled, if not it will be hard to test this in Codespaces for them (but It is still ok for Vsocde with remote extension):

https://docs.github.com/en/codespaces/managing-codespaces-for-your-organization/enabling-codespaces-for-your-organization

**Comment by bhack** [comment]
Just a small reminder. 
Codespaces is in free trial until September 10, 2021 then we need to pay the VM instance to test this on the Github Codespace VM or with the Vscode browser. 

Of course it will be still free to test on our your HW resources with Vscode desktop and the remote extension.

**Comment by bhack** [comment]
@angerson @mihaimaruseac I have minimized he Dockerfile and started to use your new `manylinux2014` (currently gpu-only) images.

This new setup is oriented to a Docker rootless installation. See
https://github.com/microsoft/vscode-remote-release/issues/4646
https://github.com/moby/moby/issues/41497

File: .devcontainer/devcontainer.json [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+		"ms-vscode.cpptools",
+	],
+
+	// Use 'forwardPorts' to make a list of ports inside the container available locally.
+	// "forwardPorts": [],
+
+	// Use 'postCreateCommand' to run commands after the container is created.
+	// "postCreateCommand": "uname -a",
+    "mounts": [
+        "source=tensorflow_cache,target=/root/.cache/bazel/"
```
Comment (bhack) [inline .devcontainer/devcontainer.json:48]: @chuxel Do you know why also using this the `bazel` cache is invalidate on every vscode restart?

**Comment by bhack** [comment]
@mihaimaruseac @angerson @yarri-oss You can test this clicking on the code button

immagine

**Comment by bhack** [comment]
Note: with the Codespace "free tier" available disk space you will have a `no space left on device` as with our manylinux/docker refactory effort we still have a CUDA embedded image only that is very large to run on the "free tier". 
See more at https://github.com/tensorflow/build/pull/47

File: .devcontainer/devcontainer.json (hunk 1 above)
Comment (bhack) [reply .devcontainer/devcontainer.json:48]: I've debugged a little more and it is seem to related `conde insider` (dality) update restart. Let me know if you want an upstream ticket.

Comment (bhack) [reply .devcontainer/devcontainer.json:55]: @Chuxel  I've an upstream ticket now at https://github.com/microsoft/vscode-dev-containers/issues/1453

**Comment by bamurtaugh** [comment]
Hi folks 👋! I wanted to provide some updates on development containers that might make this PR more interesting (cc @angerson as part of the initial PR thread).

Our team is working on an open dev container specification, so that any user in any tool can create and connect to dev containers. This includes tools beyond VS Code or Codespaces.

We recently released the reference implementation as an [open source CLI](https://github.com/devcontainers/cli), in addition to an [initial version of the spec](https://github.com/devcontainers/spec/blob/main/docs/specs/devcontainer-reference.md). We now have a [`devcontainers` org](https://github.com/devcontainers) for these repos and a site where you can read more about this effort: https://containers.dev/.

We also introduced a `customizations` property in which tool-specific properties live ([additional info](https://github.com/devcontainers/spec/issues/1)).

The CLI and spec are both in active development, so they'll continue to evolve, especially with external feedback. We'd love to hear if the spec helps make this PR more viable to become part of TensorFlow, and if there is any feedback or questions that'd make this a better fit - we're more than happy to discuss and collaborate. Thanks!

**Comment by bhack** [comment]
Thank you @bamurtaugh  for your team updates I hope that @learning-to-play and @angerson could re-evaluate this.

In the meantime, on my side, one of the major blockers on this is the upstream issue we had https://github.com/microsoft/vscode-remote-release/issues/6746.

**Comment by gbaned** [comment]
@learning-to-play, @angerson, @bhack  Any update on this PR? Please. Thank you!

**Comment by bhack** [comment]
@theadactyl Can we try to reply to @bamurtaugh She is a Vscode PM at Microsoft? 

Thanks

**Comment by bhack** [comment]
@bamurtaugh Any new on Vscode side for https://github.com/microsoft/vscode-remote-release/issues/6746?

**Comment by bamurtaugh** [comment]
Thanks for the ping @bhack! 

Our team has been actively focused on specification work and an updated plan for community contributions and the vscode-dev-containers repo moving forward, which we've now outlined in this issue: https://github.com/microsoft/vscode-dev-containers/issues/1589. This includes moving to a couple new repos for images and Features (https://github.com/devcontainers/images, https://github.com/devcontainers/features), and we anticipate to have a similar repo and process for templates/definitions.

Please let me know if you have any questions, thank you!

**Comment by bhack** [comment]
@bamurtaugh For TF we need to find a solution  for the VSCode ticket at https://github.com/microsoft/vscode-remote-release/issues/6746

Do you have any update?

**Comment by bamurtaugh** [comment]
Apologies for a potential lack of clarity in my original comment. I meant to explain that our team's primary work has been on the dev container spec and the items I highlighted. https://github.com/microsoft/vscode-remote-release/issues/6746 is on the backlog, but spec work has been at the forefront, so there are no new updates on https://github.com/microsoft/vscode-remote-release/issues/6746 at this time.

**Comment by bhack** [comment]
@bamurtaugh Thanks for the clarification.

**Comment by bhack** [comment]
Just a side note for @tensorflow/sigbuild-dockerfile-editors. 
Now the devcontainers could be tested with an official Github Action:
https://github.com/devcontainers/ci/blob/main/docs/github-action.md

**Comment by bhack** [comment]
> Just a side note for @tensorflow/sigbuild-dockerfile-editors. Now the devcontainers could be tested with an official Github Action: https://github.com/devcontainers/ci/blob/main/docs/github-action.md

I've added this with a PR in Keras-cv https://github.com/keras-team/keras-cv/pull/893

**Comment by gbaned** [comment]
Hi @bhack, @bamurtaugh, @angerson  Any update on this PR? Please. Thank you!

**Comment by bhack** [comment]
> Hi @bhack, @bamurtaugh, @angerson Any update on this PR? Please. Thank you!

Still waiting a solution from @bamurtaugh and the MS devcontainer team for https://github.com/microsoft/vscode-remote-release/issues/6746

On the TF side I need to check if @MichaelHudgins is still interested in this work.

**Comment by gbaned** [comment]
Hi @bhack  Any update on this PR? Please. Thank you!

**Comment by bhack** [comment]
Still waiting for that ticket

**Comment by gbaned** [comment]
Hi @bhack Any update on this PR? Please. Thank you!

**Pull request closed without merging**